Add userspace emulation of fork to the NaCl Internal Runtime (IRT)

alyptik opened the pull request:
Add a single-process implementation (using the Linux NPTL implementation
of POSIX Threads to simulate common multi-process programming) of the fork
system call. Programming and system designs leveraging this style are
ubiquitous in modern Computer Science and Networking; common implementation
of a standard web server is the single HTTP parent daemon which forks off
several children worker processes.

The main motivation for an emulation of fork (as opposed to just using
the kernel system call) is allowing for _much_ stricter control and
auditing of application(s) being run within NaCl while still being
transparent to sandboxed processes.

This provides users with a balance between useful security and practical
usability which, if not respected, tends to become "a system safe from
everyone; owner included."

alyptik committed:
lind: extend Lind system call interfaces

Add the following system calls:
  - pipe
  - pipe2
  - fork
  - execv
  - execve
  - getppid
  - waitpid
  - wait
  - wait4
  - sigprocmask

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
fork: add userspace emulation of fork to the NaCl Internal Runtime (IRT)

Add a single-process implementation (using the Linux NPTL implementation
of POSIX Threads to simulate common multi-process programming) of the fork
system call. Programming and system designs leveraging this style are
ubiquitous in modern Computer Science and Networking; common implementation
of a standard web server is the single HTTP parent daemon which forks off
several children worker processes.

The main motivation for an emulation of fork (as opposed to just using
the kernel system call) is allowing for _much_ stricter control and
auditing of application(s) being run within NaCl while still being
transparent to sandboxed processes. Providing users with a sane
compromise between security and practicality avoids designing security
systems to be "safe from everyone; owner included."

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
@detectivelyw Could you please review commit [9966667b6f](native_client:9966667b6f)? These are largely your Repy related changes and would appreciate a second pair of eyes to make sure I didn't remove anything I shouldn't have.

alyptik commented:
@ex0dus-0x Would you mind taking a look at the `src/trusted/service_runtime/sel_*` files? Most of those changes involve the TCB between the host system and NaCl where I know your opinions would be very helpful.

alyptik commented:
@pkmoore Would you mind taking a look at the `src/trusted/service_runtime/nacl_*` (including the `nacl_syscall_handlers_gen.py`) files please? These contain the the actual IRT system call implementation of fork, and since you have a lot of experience at the nitty-gritty system call layer your input would be a huge help.

alyptik commented:
@JustinCappos @moyix I wasn't sure how busy you two were, so I didn't want to assign anything explicity for you guys. That being said, I would still very much appreciate both of your opinions if you have time.

alyptik commented:
I very much appreciate the help, thank you guys.

ex0dus-0x commented on src/trusted/service_runtime/sel_ldr.c:
```diff
@@ -666,17 +680,17 @@ int NaClAddThread(struct NaClApp        *nap,
-    NaClLog(LOG_FATAL,
+     NaClLog(LOG_FATAL,
             "NaClRemoveThreadMu:: DynArraySet at position %d failed\n",
             thread_num);
```
@alyptik noticed some whitespacing inconsistencies here.

alyptik replied:
This change was intended; the original was indented with one space incorrectly.

ex0dus-0x commented on src/trusted/service_runtime/sel_ldr_standard.c:
```diff
@@ -904,6 +874,215 @@ int NaClCreateMainThread(struct NaClApp     *nap,
+  /* now actually spawn the thread */
+  NaClXMutexLock(&nap_child->mu);
+  nap_child->running = 1;
+  NaClXMutexUnlock(&nap_child->mu);
```
Can you explain why the critical section here is only for setting the running flag for `nap_child`?

alyptik replied:
We need a lock for two reasons:

1) To act as a memory barrier and synchronization point for when the parent NaClApp and functions like NaClReportExitStatus() try to query thread running state.

2) To block strange spurious thread spawns when running on a host with glibc 2.28 or later.

I am still trying to figure out the reason behind the spurious thread spawns, but I've had no luck yet. C11 threads *were* added in glibc 2.28 which may be related, but haven't found anything concrete so far.

alyptik replied:
For what it's worth, I did copy this lock straight from the non-fork thread constructor function, so it is possible this may not be absolutely needed here.

I mostly looked at other code which read this variable and the lock seemed sane, but I'll do some testing to verify it. Good catch!

alyptik committed:
platform: tidy up and properly comment preprocessor sections

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
@aaaaalbert Whenever you have some free time, I would appreciate it if you could take a look at the `src/trusted/service_runtime/nacl_*` (including the `nacl_syscall_handlers_gen.py`) files. These contain the the actual IRT system call implementation of fork() and are the largest part, so another pair of eyes in addition to @pkmoore is probably a good idea.

@awwad I know you don't do much C, but would also appreciate just a quick glance over those files to maybe see if there are any hugely glaring mistakes I've missed.

Thanks!

ex0dus-0x commented on src/trusted/service_runtime/posix/sel_memory.c:
```diff
@@ -56,6 +56,16 @@ static int NaClPageAllocInternalFlags(void **p, size_t size, int map_flags) {
+  if (*p) {
+      map_flags |= MAP_FIXED;
+  }
+  if (NaClPageAllocInternalFlags(p, size, map_flags) < 0) {
```
Can you explain the functionality of this function in context of your `fork()` implementation?

alyptik replied:
This is basically a wrapper for the internal function to allocate pages. The function itself, NaClPageAllocInternalFlags(), is the equivalent of mmap().

When we create a copy of the NaClApp during fork, we also need to make a copy of its address space. This function is what is used to allocate new pages when we walk the parent's page table for making copies of each page which doesn't refer to a file backed by persistent storage.

alyptik replied:
I should add that the reason we need the wrapper in the first place is because `NaClPageAllocInternalFlags()` is static.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+  if ((*mod_status = NaClAppPrepareToLaunch(nap_child)) != LOAD_OK) {
+    NaClLog(LOG_FATAL, "Failed to prepare child nap_parent for launch\n");
+  }
+  NaClLog(1, "Loading blob file %s\n", nap_child->nacl_file);
```
This (and other similar occurences of `NaClLog`) should probably use a log level macro, not a hardcoded number as its first argument.

alyptik replied:
The problem is there are 7 levels, which only are visible with that many -v flags (e.g. level 3 needs -vvv). I think adding macros for those would be a little bit too much abstraction without providing much benefit.

aaaaalbert replied:
I just found it odd that there's a `LOG_FATAL` but then `1` to `3`.

alyptik replied:
Hm, maybe there is a better way to wrap these logging functions, I'll give it some thought.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+   * __m256 objects are passed (8 mod 32 if __m256), after the call.
+   * Note the current doc (as of 2009-12-09) at
+   *
+   *   http://www.x86-64.org/documentation/abi.pdf
```
Down for me.

alyptik replied:
Ah yep, I'll remove this from the comment.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -231,6 +511,137 @@ struct NaClAppThread *NaClAppThreadMake(struct NaClApp *nap,
+
+/* examine arbitrary stack values */
+#if defined(_DEBUG)
+# define NUM_STACK_VALS 16
```
That's an interesting indent for the `#if`'s "body".

alyptik replied:
This indent was to keep everything consistent with the other parts of the codebase.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -239,17 +650,28 @@ int NaClAppThreadSpawn(struct NaClApp *nap,
   if (!NaClThreadCtor(&natp->host_thread, NaClAppThreadLauncher, (void *) natp,
                       NACL_KERN_STACK_SIZE)) {
-    /*
+     /*
```
The edit breaks alignment of `*` with the lines below.

alyptik replied:
Good catch, will fix that.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -302,9 +296,7 @@ int32_t NaClOpenAclCheck(struct NaClApp *nap,
-#define FLOG(VAR, BIT) do {\
-      NaClLog(1, "%s: %s\n", #BIT, (VAR & BIT) ? "yes" : "no");\
-    } while (0)
+#define FLOG(VAR, BIT) NaClLog(1, "%s: %s\n", #BIT, ((VAR) & (BIT)) ? "yes" : "no")
```
This `#define` feels out of place here in the middle of the code. Move up / out to group with other log functions?

alyptik replied:
Yeah, you are right, I'll hoist that out of the function.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -458,54 +442,88 @@ int32_t NaClSysNameService(struct NaClAppThread *natp,
+
+  if (nap->fd > FILE_DESC_MAX) {
+    ret = -NACL_ABI_EBADF;
+    goto done;
```
What's the benefit of the `goto` here and below, compared to a plain `return -NACL_ABI_...`?

alyptik replied:
The gotos are to keep the error path code style consistent with the rest of the syscalls that use gotos because of their more complicated error handling.

aaaaalbert commented:
One general remark: variable names like `nap` and `natp` have a slight Hungarian touch, so to speak.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -538,30 +556,54 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
   char                 path[NACL_CONFIG_PATH_MAX];
   nacl_host_stat_t     stbuf;
   int                  allowed_flags;
+  const char           glibc_prefix[] = "/lib/glibc/";
```
Is `char []` for a string preferred over `char *`?

alyptik replied:
Ah yeah, that is a good idea, will fix.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -538,30 +556,54 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
+   *
+   * -jp
+   */
+  if (!memcmp(path, tls_prefix, sizeof tls_prefix - 1)) {
```
Possible multi-byte char encoding issue ahead?

alyptik replied:
The glibc build always uses `/lib/glibc/tls` so it shouldn't be a problem.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -578,68 +620,93 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
+     nap->num_lib++;
+  }
+
+  NaClLog(1, "[*** Debug ***][Open] fd = %d, filepath = %s \n", fd_retval, path);
```
Besides the hardcoded log level mentioned previously, the debug message is quite different from the others I saw so far: custom prefix, contains only half of the function name; trailing space in message.

alyptik replied:
Whoops, missed that. Will fix.

aaaaalbert commented:
Most functions seem to log "entering ..." and inner function calls with their parameters and return values. I wonder if there's a prettier and less verbose way to achieve this.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.h:
```diff
@@ -105,21 +110,21 @@ int32_t NaClSysStat(struct NaClAppThread *natp,
 
 int32_t NaClSysMkdir(struct NaClAppThread *natp,
-                     uint32_t             path,
+                     uint32_t             pathname,
```
Does that variable name make sense? This isn't a pathname in the common meaning of the word.

alyptik replied:
This was to keep consistency with  the other system call parameter naming conventions.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.h:
```diff
@@ -301,6 +306,14 @@ int32_t NaClSysTestInfoLeak(struct NaClAppThread *natp);
+int32_t NaClSysPipe(struct NaClAppThread  *natp, uint32_t *pipedes);
+int32_t NaClSysFork(struct NaClAppThread  *natp);
+int32_t NaClSysExecv(struct NaClAppThread  *natp);
+int32_t NaClSysExecve(struct NaClAppThread  *natp, void* path, void* argv, void* envp);
```
Style consistency for pointer names (`... *natp` vs `void* ...`).

alyptik replied:
Ah, good catch. Missed this in my original revision, will fix.

aaaaalbert commented:
`main` in `src/trusted/service_runtime/nacl_syscall_handlers_gen.py` isn't in great shape, and IMO we should deprecate Python 2 wherever we can.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+  struct NaClApp *nap_parent = natp->nap;
+  NaClErrorCode *mod_status = NULL;
+
+  CHECK(nap_master);
```
What `CHECK` is performed here? (I can look it up, but perhaps the macro name can guide my understanding.)

aaaaalbert replied:
This seems to do pretty much the same, including spelling out the checked condition:

```c
#include <assert.h>
int main(int argc, char** argv) {
  assert ( argc == 0 );
  return 0;
}
```
Compile and run (OS X 10.11.6, clang-800.0.42.1):
```
Assertion failed: (argc == 0), function main, file foo.c, line 3.
Abort trap: 6
```

alyptik replied:
This is just an assert() basically, but the rest of the codebase uses CHECK() so I was keeping things consistent.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+    NaClXMutexUnlock(&nap_cur->children_mu);
+
+    /* break after parent's cleanup */
+    if (nap_cur == nap_parent) {
```
I don't get the combination of this `if ... break` and `nap_cur = nap_parent` in the head of the `for` loop. Can the conditional `break` just be rolled into the termination condition of the `for` loop?

alyptik replied:
This is just to perform the loop body operations on both parent and master. I think maybe it it possible squeeze the break condition into the for statement though, let me see if I can figure out a way.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+    NaClSetDesc(nap_child, new_fd, old_nd);
+    fd_cage_table[nap_child->cage_id][new_fd] = fd_cage_table[nap_parent->cage_id][old_fd];
+    nap_child->fd++;
+    NaClLog(1, "NaClGetDesc() copied parent fd [%d] to child fd [%d]\n", old_fd, new_fd);
```
`... [%d] ...` I think introduces yet a new debug message sub-format.

alyptik replied:
Yeah the debug format is a little consistent, I'll see if I can figure out a slightly cleaner way to do this.

aaaaalbert commented on src/trusted/service_runtime/nacl_syscall_common.c:
```diff
@@ -578,68 +620,93 @@ int32_t NaClSysOpen(struct NaClAppThread  *natp,
-              retval);
+    if (!retval) {
+      retval = NaClSetAvail(nap, ((struct NaClDesc *) NaClDescDirDescMake(hd)));
+      NaClLog(1, "Entered directory into open file table at %d\n", retval);
```
Watch out for other uses of `enter` (i.e. where entering a function) in debug messages. Perhaps reword here.

alyptik replied:
Yeah, I'll reword that.

aaaaalbert commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -16,14 +16,235 @@
+  uint32_t thread_idx;
+  nacl_reg_t secure_stack_ptr;
+
+  NaClLog(1, "%s\n", "NaClAppForkThreadLauncher: entered");
```
Inconsistent message format, compare e.g. [here](https://github.com/Lind-Project/native_client/pull/6/files#diff-4c3032ce019016f86159cf8d6d452182R62).

alyptik replied:
Same as last comment.

alyptik commented:
@aaaaalbert As much as I agree with you regarding the names of nap and natp, I've just been following Google's original naming convention for those to keep things consistent as that is how they are named across the entire codebase.

The function log thing is an interesting point, you are probably right. I'll see if I can figure out a way to do so.

I also agree with you on deprecating Python 2, that's something I plan to address in a future PR.

aaaaalbert commented:
Okay, I thought so re: `nap` and `natp`. Good to have your confirmation!

aaaaalbert commented:
Yet another general remark: Perhaps reconsider the various platform-dependent `#ifdef`s. See section 8 of Ousterhout's [Tcl/Tk Engineering Manual](http://wfr.tcl-lang.org/fichiers/Ousterhout_Engineering_Manual.pdf) for suggestions on selecting functionality differently.

alyptik commented:
@aaaaalbert That's an interesting suggestion; I'll give that a read right now, thanks!

alyptik committed:
refactoring: coalesce loop break conditions

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik committed:
linting: correct minor code style issues

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
@aaaaalbert Pushed some commits to address your suggestions; redoing the debug messages for consistency will require a bit more thought and may have to be done in a future PR, but everything else was pretty smooth. What are your thoughts?

aaaaalbert commented:
94e1a351f6be08cd140ae8b4fe8cb010cff9c860 is trivially OK, though I wonder if *some* ABI docs should be referenced instead of these.

21089dae11e0f629d7b342958ec4edc64ede26c3 also looks good.

1eb5e496dc064221f3a56a32ac9544c9d7b6dc98 looks much cleaner for the removed `break`, thanks! I didn't check in great detail the other changes that this commit introduced; not sure if they are only cleanup / fallout from rewriting the loop conditions.

alyptik commented:
@aaaaalbert Ah, I'll see if I can find some ABI docs and will rebase if so. And regarding the break, it was indeed just that and a rename from nap_cur to cur to avoid huge lines.

alyptik committed:
documentation: add x86-64 ABI docs

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
@aaaaalbert Found a copy :)

aaaaalbert commented:
:+1:

detectivelyw commented on src/shared/platform/lind_platform.c:
```diff
@@ -5,26 +5,33 @@
+
+static int initialized;
 
 #define REPY_RELPATH "../repy/"
```
Does this path definition require corresponding file directory to be setup during installation? If so, perhaps we should provide comments and documentation to make sure things are setup correctly.

alyptik replied:
Yeah, the macro just obfuscates this part, as it's really only used in one place. Just going to remove it and use a string literal.

detectivelyw commented on src/shared/platform/lind_platform.c:
```diff
@@ -240,89 +249,94 @@ int ParseResponse(PyObject* response, int* isError, int* code, char** dataOrMess
+        }
 
-#define DUMP_DATA(x) printf(#x" = 0x%"NACL_PRIX64"\n", (uint64_t)(x));
+#define DUMP_DATA(x) printf(#x" = %#lx\n", (uint64_t)(x))
```
Is there a reason for changing the NaCl default data format?

alyptik replied:
After some examination, it turns this macro isn't even useful, as it's only used in an `#if 0` section.  Removing the dead code.

detectivelyw commented on src/shared/platform/lind_platform.c:
```diff
@@ -240,89 +249,94 @@ int ParseResponse(PyObject* response, int* isError, int* code, char** dataOrMess
+            DUMP_DATA((x)->st_size);                                   \
+            DUMP_DATA((x)->st_blksize);                                \
+            DUMP_DATA((x)->st_blocks);                                 \
+        } while (0)
```
What's happening here?

alyptik replied:
See previous resolution.

detectivelyw commented on src/trusted/service_runtime/lind_syscalls.c:
```diff
@@ -259,46 +268,50 @@ int LindSelectPreprocess(struct NaClApp *nap, uint32_t inNum, LindArg* inArgs, v
-    NaClLog(3, "max_fd is set to %"NACL_PRId64" was %"NACL_PRId64"\n", *(int64_t*)&inArgs[0].ptr, max_fd);
+    *(int64_t *)&inArgs[0].ptr = max_hfd + 1;
+    ((int *)(*xchangedata))[0] = max_hfd + 1;
+    NaClLog(1, "max_fd is set to %"NACL_PRId64" was %"NACL_PRId64"\n", *(int64_t *)&inArgs[0].ptr, max_fd);
```
So here we are using the NaCl defined data format NACL_PRId64. I think it would be nice to keep consistence about our NaClLog data format. (See an inconsistent case before.)

alyptik replied:
Yes, agreed. This format is probably best, as it's already established in many places.

detectivelyw commented on src/trusted/service_runtime/lind_syscalls.c:
```diff
@@ -720,95 +768,105 @@ static int NaClCopyZStr(struct NaClApp *nap,
+     *        and returns two fds for the pipe
+     */
+    if (callNum == 61) {
+        int data[2] = {9001, 9002};
```
So this is my old testing code, which used hardcoded 9001 and 9002 as fds for our own pipe. If we are using different fds for our pipes now, this should be modified. If we are not using this code now, I think we should put this piece of code in a comment and mark it as "unused".

alyptik replied:
This has been removed as per @JustinCappos in my other branches, but it does seem better to just not add it at all, yeah. Going to just rip it out (of course with testing to be sure this doesn't break anything).

JustinCappos commented:
Or just delete it.  No need to junk up the code base with obsolete,
commented code when we can pull it out of github later if we really need it.

On Tue, Sep 25, 2018 at 5:32 PM Yiwen Li <notifications@github.com> wrote:

> *@detectivelyw* commented on this pull request.
> ------------------------------
>
> In src/trusted/service_runtime/lind_syscalls.c
> <https://github.com/Lind-Project/native_client/pull/6#discussion_r220363120>
> :
>
> >      }
>
> -    /*
> -    // For debugging
> -    for(int i=0; i<(int)inNum; ++i) {
> -        DumpArg(&inArgSys[i]);
> +    /* yiwen: handle lind_pipe here.
> +     *        the pipe() call should initialize the pipe buffer
> +     *        and returns two fds for the pipe
> +     */
> +    if (callNum == 61) {
> +        int data[2] = {9001, 9002};
>
> So this is my old testing code, which used hardcoded 9001 and 9002 as fds
> for our own pipe. If we are using different fds for our pipes now, this
> should be modified. If we are not using this code now, I think we should
> put this piece of code in a comment and mark it as "unused".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Lind-Project/native_client/pull/6#pullrequestreview-158754905>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA0XDwdvqoERKzUTW8-2QhcrQB9lt9E8ks5ueqECgaJpZM4WPB_2>
> .
>

detectivelyw commented on src/shared/platform/lind_platform.c:
```diff
@@ -240,89 +249,94 @@ int ParseResponse(PyObject* response, int* isError, int* code, char** dataOrMess
+        } while (0)
 #else
-#define DUMP_STAT(x)
+# define DUMP_STAT(x) do {/* no-op */} while (0)
```
What's this?

alyptik replied:
See previous resolution.

alyptik committed:
platform: remove unused macros

The REPY_RELPATH macro is only used in one code section and doesn't
convey any useful information, and the DUMP_DATA() and DUMP_STAT()
macros are old commented out code.

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
@detectivelyw @JustinCappos So thinking about this more, most of the problems that we've been finding are largely because at the start of all this I based most of the work off of caging and ended up just copying the Repy parts (in the beginning nothing seemed to work if I based my branches off of lind).

Now that it is starting to come together, mostly due to explanations of Repyv2 from @aaaaalbert, there seems to be a large amount of bitrot leftover from the original SVN imports, which seem to be causing a lot of code style issues. Things like old commented out code, unneeded macros, and oddly named variables are scattered about these sections.

So at this point I am wondering if you guys think it would be useful to rebuild the Repy and SeattleLib parts cleanly from scratch?

Since these issues are only style-related, as everything does work, I'm not sure if a major cleanup like this is useful to think about now, or maybe if we should just shelf it for later?

alyptik committed:
platform: remove commented-out dead code

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

detectivelyw commented:
> @detectivelyw @JustinCappos So thinking about this more, most of the problems that we've been finding are largely because at the start of all this I based most of the work off of caging and ended up just copying the Repy parts (in the beginning nothing seemed to work if I based my branches off of lind).
> 
> Now that it is starting to come together, mostly due to explanations of Repyv2 from @aaaaalbert, there seems to be a large amount of bitrot leftover from the original SVN imports, which seem to be causing a lot of code style issues. Things like old commented out code, unneeded macros, and oddly named variables are scattered about these sections.
> 
> So at this point I am wondering if you guys think it would be useful to rebuild the Repy and SeattleLib parts cleanly from scratch?
> 
> Since these issues are only style-related, as everything does work, I'm not sure if a major cleanup like this is useful to think about now, or maybe if we should just shelf it for later?

I vote for shelving a major cleanup for later. 
Our codebase works now, and I think most code has good / consistent style. So I don't think it is necessary to rebuild from scratch. And I think fixing style-related issues as we keep building on the existing codebase is a reasonable and practical approach.

detectivelyw commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -32,7 +253,7 @@ void WINAPI NaClAppThreadLauncher(void *state) {
-  NaClLog(4, "      natp = 0x%016"NACL_PRIxPTR"\n", (uintptr_t) natp);
+  NaClLog(4, "     natp  = 0x%016"NACL_PRIxPTR"\n", (uintptr_t)natp);
   NaClLog(4, " prog_ctr  = 0x%016"NACL_PRIxNACL_REG"\n", natp->user.prog_ctr);
   NaClLog(4, "stack_ptr  = 0x%016"NACL_PRIxPTR"\n",
```
Is there any reason for us to use "4" here in the NaClLog? Other messages are directed to the standard output "1".

alyptik replied:
This particular log message is fairly noisy and not usually useful information, so I have made it require -vvvv to be visible.

detectivelyw commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -147,13 +426,12 @@ void NaClAppThreadTeardown(struct NaClAppThread *natp) {
   NaClThreadExit();
-  NaClLog(LOG_FATAL,
-          "NaClAppThreadTeardown: NaClThreadExit() should not return\n");
+  NaClLog(LOG_FATAL, "NaClAppThreadTeardown: NaClThreadExit() should not return\n");
```
So here we use "LOG_FATAL", but at other places we were using numbers like 1, 2, 3, 4. I wonder if we could make it consistent and easy to read / understand.

alyptik replied:
It's a bit odd because the number refers to number of -v needed to see the message, with macros for the special cases. I am aware it's not ideal, and I'm still trying to figure out of there's a better way. This would definitely be a change for a seperate PR, though.

alyptik commented:
@detectivelyw Yeah, I am sort of leaning that way as well. Also interested in your thoughts, @aaaaalbert and @moyix

detectivelyw commented on src/trusted/service_runtime/nacl_app_thread.c:
```diff
@@ -81,25 +81,20 @@ struct NaClApp *NaClChildNapCtor(struct NaClAppThread *natp) {
-    NaClXMutexLock(&nap_cur->children_mu);
-    if (nap_cur->children_ids[nap_cur->num_children]) {
-       nap_cur->num_children++;
+  for (struct NaClApp *cur = nap_master; cur == nap_master || cur == nap_parent; cur = nap_parent) {
```
In my opinion, nap_cur seems like a better name than cur, since it tells us more about what data structure it represents.

alyptik replied:
Yeah you are right, will revert the rename part here.

detectivelyw commented on src/trusted/service_runtime/sel_ldr.c:
```diff
@@ -57,8 +59,28 @@
+double time_counter = 0.0;
+double time_start = 0.0;
+double time_end = 0.0;
+int cage = 1001;
```
I believe this (cage = 1001) isn't used any more. Please help remove it.

alyptik replied:
Indeed it's not, will remove.

aaaaalbert commented:
So, @detectivelyw and @alyptik, you suggest:
> I vote for shelving a major cleanup for later.

I don't think this would be a major cleanup actually. As far as I can see, none of the difficult things you are trying to achieve with NaCl/Repy are more difficult with an up-to-date version of repy_v2 / seattlelib_v2. IMHO you'd get lots of benefits from updating.

* The build process isn't based off `preparetest.py` anymore, see [the new build docs](https://github.com/SeattleTestbed/docs/blob/master/Contributing/BuildInstructions.md).
* Through this, the codebase is laid out in a more meaningful way. We've split the subdirs of our old SVN (i.e. what you have here!) into indepedent repos.
* This makes it more straightforward to select and use (and modify) only the parts that you are interested in, e.g. *not* the nodemanager or experimenter shell.
* The preprocessing step (from `.mix` to `.repy`) is replaced by a more Python-like import process using `dylink`.

That's what I can think off right now, perhaps there are other things that I don't see yet where it matters more.

alyptik commented:
@aaaaalbert After talking with @JustinCappos it actually seems most of the code is v2, just not the most recent version of it. It shouldn't be too much work to clean up and will be giving it a go in a test branch, thanks for the feedback on this.

alyptik committed:
linting: revert nal_cur name change and remove unused cage variable

Signed-off-by: Joey Pabalinas <joeypabalinas@gmail.com>

alyptik commented:
The feedback was all very helpful guys, thanks! I've addressed what was feasible and am merging it in.

aaaaalbert commented:
> After talking with @JustinCappos it actually seems most of the code is v2, just not the most recent version of it.

Exactly, the single biggest difference is in the dir/repo layout and build scripts.
